src/scheduler/batch/main.ts

```diff
@@ -0,0 +1,90 @@
+import { List } from "server/models/types";
+import { logger } from "server/services/logger";
+import { findListItems } from "server/models/listItem";
+import { SCHEDULED_PROCESS_TODAY_DATE } from "server/config";
+import * as helpers from "./helpers";
+import { getCurrentAnnualReviewData, getDateForContext } from "./helpers";
+import { ListItemWithHistory } from "server/components/dashboard/listsItems/types";
+import { formatDate } from "../helpers";
+import { isBefore, subDays, subMonths } from "date-fns";
+
+export async function populateCurrentAnnualReview(
+  lists: List[],
+  contexts: helpers.SchedulerDateContexts
+): Promise<void> {
+  const listIds = lists.map((list) => list.id);
+  const findListItemsResult = await findListItems({
+    listIds,
+    statuses: ["PUBLISHED", "CHECK_ANNUAL_REVIEW"],
+    isAnnualReview: false,
+  });
+  if (findListItemsResult.error) {
+    logger.error(`Unable to retrieve List Items for Lists ${listIds}: ${findListItemsResult.error.message}`);
+    return;
+  }
+  if (!findListItemsResult.result.length) {
+    logger.info(`No List Items found for Lists ${listIds}`);
+    return;
+  }
+  const listItemsForAllLists: ListItemWithHistory[] = findListItemsResult.result;
+  for (const list of lists) {
+    // exclude list items published in the last month from annual review
+    const listItemsEligibleForAnnualReview = listItemsForAllLists.filter((listItem) => {
+      const publishedEvent = listItem.history.find((event) => event.type === "PUBLISHED");
+      const oneMonthEarlier = subMonths(Date.now(), 1);
+      if (!publishedEvent?.time || isBefore(publishedEvent?.time, oneMonthEarlier)) {
+        return listItem.listId === list.id;
+      }
+      return false;
+    });
+    if (!listItemsEligibleForAnnualReview.length) {
+      logger.info(`No list items identified for list ${list.id}, excluding from sending annual review emails`);
+      return;
+    }
+    const listItemIdsForAnnualReview = listItemsEligibleForAnnualReview.map((listItem) => listItem.id);
+    const currentAnnualReview = getCurrentAnnualReviewData(listItemIdsForAnnualReview, contexts);
+    await updateListForAnnualReview(list, { currentAnnualReview });
+  }
+}
+
+export async function updateListsForAnnualReview(todayDateString: string): Promise<void> {
+  const contexts = helpers.getDateContexts(todayDateString);
+  const annualReviewStartContext = getDateForContext(
+    contexts,
+    "annualReview",
+    helpers.schedulerMilestoneDays.both.START
+  );
+  if (annualReviewStartContext) {
+    // @todo also get the lists without an annual review date that were first published 1 year ago. Only needed if List.annualReviewStartDate not populated
+    const { result } = await findListByAnnualReviewDate(annualReviewStartContext.eventDate);
+
+    // exclude lists that already have currentAnnualReview populated
+    const lists = result?.filter(list => !list.jsonData.currentAnnualReview);
```

Comment:
can this be done in the query?

Reply:
doesn't look like Prisma can handle queries to detect where a nested json object field doesn't exist so it needs to be done after the query is executed